cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -0,0 +1,381 @@
+            }
+          }
+        };
+      }
+    };
+  }
+
+  /**
+   * Returns a {@link BatchWritable} that is implemented by the given dataset. The {@link BatchWritable} returned
+   * is also {@link Closeable}. Caller of this method is responsible for calling {@link Closeable#close()}} of
+   * the returned object. This method is expected to be called by {@link SparkDatasetOutputFormat}.
+   *
+   * @param datasetName name of the dataset
+   * @param arguments arguments for the dataset
+   * @param <K> key type for the BatchWritable
+   * @param <V> value type for the BatchWritable
+   * @return the CloseableBatchWritable
+   * @throws DatasetInstantiationException if cannot find the dataset
+   * @throws IllegalArgumentException if the dataset does not implement BatchWritable
+   */
+  public <K, V> CloseableBatchWritable<K, V> getBatchWritable(String datasetName, Map<String, String> arguments) {
+    final Dataset dataset = createDataset(datasetName, arguments);
+    Preconditions.checkArgument(dataset instanceof BatchWritable, "Dataset %s of type %s does not implements %s",
+                                datasetName, dataset.getClass().getName(), BatchWritable.class.getName());
+
+    @SuppressWarnings("unchecked")
+    final BatchWritable<K, V> delegate = (BatchWritable<K, V>) dataset;
+    return new CloseableBatchWritable<K, V>() {
+      @Override
+      public void write(K key, V value) {
+        delegate.write(key, value);
+      }
+
+      @Override
+      public void close() throws IOException {
+        commitAndClose(dataset);
+      }
+    };
+  }
+
+  /**
+   * Sets the {@link ApacheSparkContext} for this context.
+   */
+  synchronized void setApacheSparkContext(ApacheSparkContext apacheSparkContext) {
+    this.apacheSparkContext = apacheSparkContext;
+    if (stopped) {
+      apacheSparkContext.stop();
+    }
+  }
+
+  /**
+   * Returns the transaction for the Spark program execution.
+   */
+  public Transaction getTransaction() {
+    return transaction;
+  }
+
+  /**
+   * Returns the {@link ApacheSparkContext}.
+   *
+   * @throws IllegalStateException if the {@link ApacheSparkContext} is not yet set.
+   */
+  private synchronized ApacheSparkContext getApacheSparkContext() {
+    Preconditions.checkState(apacheSparkContext != null,
+                             "Spark Context not available in the current execution context");
+    return apacheSparkContext;
+  }
+
+  /**
+   * Calls {@link TransactionAware#commitTx()} on the given dataset followed by a {@link Closeable#close()} call.
+   */
+  private void commitAndClose(Dataset dataset) {
+    try {
+      if (dataset instanceof TransactionAware) {
+        try {
+          ((TransactionAware) dataset).commitTx();
+        } catch (Exception e) {
+          // Nothing much we can do besides propagating it.
+          // Since the propagation will end up in the Spark Framework, we do the logging here.
+          LOG.error("Failed to commit dataset changes", e);
```

Comment:
it would be good to log the name of the dataset here. I know that Dataset does not have a getName() method. But since all callers of this method know the dataset name, can you add a parameter for that?